File: src/android/plugin/google/maps/PluginPolyline.java
package plugin.google.maps;

import com.google.android.libraries.maps.GoogleMap;
import com.google.android.libraries.maps.model.LatLng;
import com.google.android.libraries.maps.model.LatLngBounds;
import com.google.android.libraries.maps.model.Polyline;
import com.google.android.libraries.maps.model.PolylineOptions;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class PluginPolyline extends MyPlugin implements IOverlayPlugin {

  private PluginMap pluginMap;
  public final ConcurrentHashMap<String, MetaPolyline> objects = new ConcurrentHashMap<String, MetaPolyline>();

  public PluginMap getMapInstance(String mapId) {
    return (PluginMap) CordovaGoogleMaps.viewPlugins.get(mapId);
  }
  public PluginPolyline getInstance(String mapId) {
    PluginMap mapInstance = getMapInstance(mapId);
    return (PluginPolyline) mapInstance.plugins.get(String.format("%s-polyline", mapId));
  }
  @Override
  public void setPluginMap(PluginMap pluginMap) {
    this.pluginMap = pluginMap;
  }


  /**
   * Create polyline
   */
  public void create(final JSONArray args, final CallbackContext callbackContext) throws JSONException {

    final PolylineOptions polylineOptions = new PolylineOptions();
    int color;
    final LatLngBounds.Builder builder = new LatLngBounds.Builder();
    final JSONObject properties = new JSONObject();

    JSONObject opts = args.getJSONObject(2);
    final String hashCode = args.getString(3);

    final String polylineId = "polyline_" + hashCode;
    final MetaPolyline meta = new MetaPolyline(polylineId);

    int pointCnt = 0;
    if (opts.has("points")) {
      JSONArray points = opts.getJSONArray("points");
      List<LatLng> path = PluginUtil.JSONArray2LatLngList(points);
      for (int i = 0; i < path.size(); i++) {
        polylineOptions.add(path.get(i));
        builder.include(path.get(i));
      }
      pointCnt = path.size();
    }
    final int finalPointCnt = pointCnt;
    if (opts.has("color")) {
      color = PluginUtil.parsePluginColor(opts.getJSONArray("color"));
      polylineOptions.color(color);
    }
    if (opts.has("width")) {
      polylineOptions.width((float)(opts.getDouble("width") * density));
    }
    if (opts.has("visible")) {
      meta.isVisible = opts.getBoolean("visible");
      polylineOptions.visible(meta.isVisible);
    }
    if (opts.has("geodesic")) {
      polylineOptions.geodesic(opts.getBoolean("geodesic"));
    }
    if (opts.has("zIndex")) {
      polylineOptions.zIndex(opts.getInt("zIndex"));
    }
    if (opts.has("clickable")) {
      meta.isClickable = opts.getBoolean("clickable");
      properties.put("isClickable", meta.isClickable);
    } else {
      meta.isClickable = true;
      properties.put("isClickable", true);
    }
    properties.put("isVisible", meta.isVisible);

    // Since this plugin provide own click detection,
    // disable default clickable feature.
    polylineOptions.clickable(false);

    meta.properties = properties;
    objects.put(polylineId, meta);

    activity.runOnUiThread(new Runnable() {
      @Override
      public void run() {

        Polyline polyline = pluginMap.getGoogleMap().addPolyline(polylineOptions);
        polyline.setTag(polylineId);
        meta.polyline = polyline;

        if (finalPointCnt > 0) {
          meta.bounds = builder.build();
        } else {
          meta.bounds = new LatLngBounds(new LatLng(360, 360), new LatLng(360, 360));
        }
      }
    });

    try {
      JSONObject result = new JSONObject();
      result.put("hashCode", hashCode);
      result.put("__pgmId", polylineId);
      callbackContext.success(result);
    } catch (JSONException e) {
      e.printStackTrace();
      callbackContext.error("" + e.getMessage());
    }
  }


  @Override
  public void onDestroy() {
    super.onDestroy();
    objects.clear();
  }

  /**
   * set color
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setStrokeColor(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    int color = PluginUtil.parsePluginColor(args.getJSONArray(2));
    meta.polyline.setColor(color);;
    callbackContext.success();
  }

  /**
   * set width
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setStrokeWidth(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    float width = (float)(args.getDouble(2) * density);
    meta.polyline.setWidth(width);
    callbackContext.success();
  }

  /**
   * set z-index
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setZIndex(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    float zIndex = (float) args.getDouble(2);
    meta.polyline.setZIndex(zIndex);
    callbackContext.success();
  }


  /**
   * Remove the polyline
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void remove(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);

    MetaPolyline meta = instance.objects.remove(polygonId);
    meta.polyline.remove();
    callbackContext.success();
  }

  @PgmPluginMethod(runOnUiThread = true)
  public void setPoints(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    JSONArray positionList = args.getJSONArray(2);

    List<LatLng> path = meta.polyline.getPoints();
    path.clear();
    JSONObject position;
    for (int i = 0; i < positionList.length(); i++) {
      position = positionList.getJSONObject(i);
      path.add(new LatLng(position.getDouble("lat"), position.getDouble("lng")));
    }
    meta.polyline.setPoints(path);
    meta.bounds = PluginUtil.getBoundsFromPath(path);
    callbackContext.success();
  }

  @PgmPluginMethod(runOnUiThread = true)
  public void removePointAt(final JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    int index = args.getInt(2);

    List<LatLng> path = meta.polyline.getPoints();
    if (path.size() > index) {
      path.remove(index);
      if (path.size() > 0) {
        meta.bounds = PluginUtil.getBoundsFromPath(path);
      } else {
        meta.bounds = new LatLngBounds(new LatLng(360, 360), new LatLng(360, 360));
      }

      meta.polyline.setPoints(path);
    }
    callbackContext.success();
  }

  @PgmPluginMethod(runOnUiThread = true)
  public void insertPointAt(final JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    int index = args.getInt(2);
    JSONObject position = args.getJSONObject(3);
    final LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));

    List<LatLng> path = meta.polyline.getPoints();
    if (path.size() >= index) {
      path.add(index, latLng);
      meta.polyline.setPoints(path);
      meta.bounds = PluginUtil.getBoundsFromPath(path);
    }
    callbackContext.success();
  }

  @PgmPluginMethod(runOnUiThread = true)
  public void setPointAt(final JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    final int index = args.getInt(2);
    JSONObject position = args.getJSONObject(3);
    final LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));


    List<LatLng> path = meta.polyline.getPoints();
    if (path.size() > index) {
      path.set(index, latLng);

      // Recalculate the polygon bounds
      meta.bounds = PluginUtil.getBoundsFromPath(path);

      meta.polyline.setPoints(path);
    }
    callbackContext.success();
  }

  /**
   * set geodesic
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setGeodesic(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    boolean isGeodisic = args.getBoolean(2);
    meta.polyline.setGeodesic(isGeodisic);
    callbackContext.success();
  }

  /**
   * Set visibility for the object
   */
  @PgmPluginMethod(runOnUiThread = true)
  public void setVisible(JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    boolean isVisible = args.getBoolean(2);
    meta.polyline.setVisible(isVisible);
    meta.isVisible = isVisible;
    meta.properties.put("isVisible", isVisible);
    callbackContext.success();
  }

  /**
   * Set clickable for the object
   */
  @PgmPluginMethod
  public void setClickable(JSONArray args, CallbackContext callbackContext) throws JSONException {
    String mapId = args.getString(0);
    String polygonId = args.getString(1);
    PluginPolyline instance = getInstance(mapId);
    MetaPolyline meta = instance.objects.get(polygonId);

    boolean clickable = args.getBoolean(2);
    meta.isClickable = clickable;
    meta.properties.put("isClickable", clickable);
    callbackContext.success();
  }
}

File: src/ios/GoogleMaps/PluginPolyline.m
//
//  PluginPolyline.m
//  cordova-googlemaps-plugin v2
//
//  Created by masashi.
//
//

#import "PluginPolyline.h"

@implementation PluginPolyline

-(void)setPluginViewController:(PluginViewController *)viewCtrl
{
  self.mapCtrl = (PluginMapViewController *)viewCtrl;
}

- (void)pluginInitialize
{
  if (self.initialized) {
    return;
  }
  self.initialized = YES;
  [super pluginInitialize];
  // Initialize this plugin
}

- (void)pluginUnload
{

  // Plugin destroy
  NSArray *keys = [self.mapCtrl.objects allKeys];
  NSString *key;
  for (int i = 0; i < [keys count]; i++) {
      key = [keys objectAtIndex:i];
      if ([key hasPrefix:@"polyline_property"]) {
        key = [key stringByReplacingOccurrencesOfString:@"_property" withString:@""];
        GMSPolyline *polyline = (GMSPolyline *)[self.mapCtrl.objects objectForKey:key];
        polyline.map = nil;
        polyline = nil;
      }
      [self.mapCtrl.objects removeObjectForKey:key];
  }

  key = nil;
  keys = nil;
  
  NSString *pluginId = [NSString stringWithFormat:@"%@-polyline", self.mapCtrl.overlayId];
  [self.mapCtrl.plugins removeObjectForKey:pluginId];
}

-(PluginPolyline *)_getInstance: (NSString *)mapId {
  NSString *pluginId = [NSString stringWithFormat:@"%@-polyline", mapId];
  PluginMap *mapInstance = [CordovaGoogleMaps getViewPlugin:mapId];
  return [mapInstance.mapCtrl.plugins objectForKey:pluginId];
}

-(void)create:(CDVInvokedUrlCommand *)command
{


  NSDictionary *json = [command.arguments objectAtIndex:2];
  NSString *idBase = [command.arguments objectAtIndex:3];
  GMSMutablePath *mutablePath = [GMSMutablePath path];

  NSArray *points = [json objectForKey:@"points"];
  int i = 0;
  NSDictionary *latLng;
  for (i = 0; i < points.count; i++) {
      latLng = [points objectAtIndex:i];
      [mutablePath
        addCoordinate:
          CLLocationCoordinate2DMake([[latLng objectForKey:@"lat"] doubleValue], [[latLng objectForKey:@"lng"] doubleValue])];
  }

  dispatch_async(dispatch_get_main_queue(), ^{

      // Create the Polyline, and assign it to the map.
      GMSPolyline *polyline = [GMSPolyline polylineWithPath:mutablePath];


      BOOL isVisible = YES;
      // Visible property
      NSString *visibleValue = [NSString stringWithFormat:@"%@",  json[@"visible"]];
      if ([@"0" isEqualToString:visibleValue]) {
        // false
        isVisible = NO;
        polyline.map = nil;
      } else {
        // true or default
        polyline.map = self.mapCtrl.map;
      }
      BOOL isClickable = NO;
      if ([[json valueForKey:@"clickable"] boolValue]) {
        isClickable = YES;
      }
      if ([[json valueForKey:@"geodesic"] boolValue]) {
        polyline.geodesic = YES;
      }
      NSArray *rgbColor = [json valueForKey:@"color"];
      polyline.strokeColor = [rgbColor parsePluginColor];

      polyline.strokeWidth = [[json valueForKey:@"width"] floatValue];
      polyline.zIndex = [[json valueForKey:@"zIndex"] floatValue];

      // Since this plugin provide own click detection,
      // disable default clickable feature.
      polyline.tappable = NO;

      NSString *id = [NSString stringWithFormat:@"polyline_%@", idBase];
      [self.mapCtrl.objects setObject:polyline forKey: id];
      polyline.title = id;

      // Run the below code on background thread.
      [self.mapCtrl.executeQueue addOperationWithBlock:^{

          //---------------------------
          // Result for JS
          //---------------------------
          NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
          [result setObject:id forKey:@"__pgmId"];

          //---------------------------
          // Keep the properties
          //---------------------------
          NSString *propertyId = [NSString stringWithFormat:@"polyline_property_%@", idBase];

          // points
          NSMutableDictionary *properties = [[NSMutableDictionary alloc] init];
          [properties setObject:mutablePath forKey:@"mutablePath"];
          // bounds (pre-calculate for click detection)
          [properties setObject:[[GMSCoordinateBounds alloc] initWithPath:mutablePath] forKey:@"bounds"];
          // isVisible
          [properties setObject:[NSNumber numberWithBool:isVisible] forKey:@"isVisible"];
          // isClickable
          [properties setObject:[NSNumber numberWithBool:isClickable] forKey:@"isClickable"];
          // geodesic
          [properties setObject:[NSNumber numberWithBool:polyline.geodesic] forKey:@"geodesic"];
          // zIndex
          [properties setObject:[NSNumber numberWithFloat:polyline.zIndex] forKey:@"zIndex"];
          [self.mapCtrl.objects setObject:properties forKey:propertyId];

          CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];
          [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
      }];
  });
}



-(void)removePointAt:(CDVInvokedUrlCommand *)command
{
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{

      NSString *polylineKey = [command.arguments objectAtIndex:1];
      NSInteger index = [[command.arguments objectAtIndex:2] integerValue];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];

      // Get properties
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];

      GMSMutablePath *mutablePath = (GMSMutablePath *)[properties objectForKey:@"mutablePath"];

      [mutablePath removeCoordinateAtIndex:index];

      // update the property
      [properties setObject:mutablePath forKey:@"mutablePath"];
      [properties setObject:[[GMSCoordinateBounds alloc] initWithPath:mutablePath] forKey:@"bounds"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setPath:mutablePath];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];

  }];

}
-(void)setPoints:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{

      NSString *polylineKey = [command.arguments objectAtIndex:1];
      NSArray *positionList = [command.arguments objectAtIndex:2];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];

      // Get properties
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];

      GMSMutablePath *mutablePath = (GMSMutablePath *)[properties objectForKey:@"mutablePath"];
      [mutablePath removeAllCoordinates];

      CLLocationCoordinate2D position;
      NSDictionary *latLng;
      for (int i = 0; i < positionList.count; i++) {
          latLng = [positionList objectAtIndex:i];
          position = CLLocationCoordinate2DMake([[latLng objectForKey:@"lat"] doubleValue], [[latLng objectForKey:@"lng"] doubleValue]);
          [mutablePath addCoordinate:position];
      }

      // update the property
      [properties setObject:mutablePath forKey:@"mutablePath"];
      [properties setObject:[[GMSCoordinateBounds alloc] initWithPath:mutablePath] forKey:@"bounds"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setPath:mutablePath];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];

}

-(void)insertPointAt:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{

      NSString *polylineKey = [command.arguments objectAtIndex:1];
      NSInteger index = [[command.arguments objectAtIndex:2] integerValue];
      NSDictionary *latLng = [command.arguments objectAtIndex:3];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];

      // Get properties
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];

      GMSMutablePath *mutablePath = (GMSMutablePath *)[properties objectForKey:@"mutablePath"];

      CLLocationCoordinate2D position = CLLocationCoordinate2DMake([[latLng objectForKey:@"lat"] doubleValue], [[latLng objectForKey:@"lng"] doubleValue]);
      [mutablePath insertCoordinate:position atIndex:index];

      // update the property
      [properties setObject:mutablePath forKey:@"mutablePath"];
      [properties setObject:[[GMSCoordinateBounds alloc] initWithPath:mutablePath] forKey:@"bounds"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setPath:mutablePath];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];

}

-(void)setPointAt:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{

      NSString *polylineKey = [command.arguments objectAtIndex:1];
      NSInteger index = [[command.arguments objectAtIndex:2] integerValue];
      NSDictionary *latLng = [command.arguments objectAtIndex:3];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];

      // Get properties
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];

      GMSMutablePath *mutablePath = (GMSMutablePath *)[properties objectForKey:@"mutablePath"];

      CLLocationCoordinate2D position = CLLocationCoordinate2DMake([[latLng objectForKey:@"lat"] doubleValue], [[latLng objectForKey:@"lng"] doubleValue]);
      [mutablePath replaceCoordinateAtIndex:index withCoordinate:position];

      // update the property
      [properties setObject:mutablePath forKey:@"mutablePath"];
      [properties setObject:[[GMSCoordinateBounds alloc] initWithPath:mutablePath] forKey:@"bounds"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setPath:mutablePath];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];

}

/**
 * Set color
 */
-(void)setStrokeColor:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
      NSString *polylineKey = [command.arguments objectAtIndex:1];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];

      NSArray *rgbColor = [command.arguments objectAtIndex:2];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setStrokeColor:[rgbColor parsePluginColor]];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];

}

/**
 * Set width
 */
-(void)setStrokeWidth:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
      NSString *polylineKey = [command.arguments objectAtIndex:1];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];
      float width = [[command.arguments objectAtIndex:2] floatValue];

      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setStrokeWidth:width];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];
}

/**
 * Set z-index
 */
-(void)setZIndex:(CDVInvokedUrlCommand *)command
{
  
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
      NSString *polylineKey = [command.arguments objectAtIndex:1];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];
      int zIndex = [[command.arguments objectAtIndex:2] intValue];

      // Update the property
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];
      [properties setObject:[NSNumber numberWithInt:zIndex] forKey:@"zIndex"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      // Run on the UI thread
      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setZIndex:(int)zIndex];
      }];


      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];
}

/**
 * Set clickable
 */
-(void)setClickable:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{

      NSString *polylineKey = [command.arguments objectAtIndex:1];
      //GMSPolyline *polyline = (GMSPolyline *)[self.mapCtrl.objects objectForKey:polylineKey];
      Boolean isClickable = [[command.arguments objectAtIndex:2] boolValue];

      // Update the property
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];
      [properties setObject:[NSNumber numberWithBool:isClickable] forKey:@"isClickable"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];
}
/**
 * Set visibility
 */
-(void)setVisible:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
      NSString *polylineKey = [command.arguments objectAtIndex:1];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];
      Boolean isVisible = [[command.arguments objectAtIndex:2] boolValue];

      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];
      [properties setObject:[NSNumber numberWithBool:isVisible] forKey:@"isVisible"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];


      // Run on the UI thread
      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          if (isVisible) {
            polyline.map = self.mapCtrl.map;
          } else {
            polyline.map = nil;
          }
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];
}
/**
 * Set geodesic
 */
-(void)setGeodesic:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
      NSString *polylineKey = [command.arguments objectAtIndex:1];
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];
      Boolean isGeodisic = [[command.arguments objectAtIndex:2] boolValue];

      // Update the property
      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      NSMutableDictionary *properties = [NSMutableDictionary dictionaryWithDictionary:
                                         [polylineInstance.mapCtrl.objects objectForKey:propertyId]];
      [properties setObject:[NSNumber numberWithBool:isGeodisic] forKey:@"isGeodisic"];
      [polylineInstance.mapCtrl.objects setObject:properties forKey:propertyId];


      // Run on the UI thread
      [[NSOperationQueue mainQueue] addOperationWithBlock:^{
          [polyline setGeodesic:isGeodisic];
      }];

      CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
      [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];
}

/**
 * Remove the polyline
 */
-(void)remove:(CDVInvokedUrlCommand *)command
{
  
  NSString *mapId = [command.arguments objectAtIndex:0];
  PluginPolyline *polylineInstance = [self _getInstance:mapId];
  
  [polylineInstance.mapCtrl.executeQueue addOperationWithBlock:^{
    NSString *polylineKey = [command.arguments objectAtIndex:1];
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
      GMSPolyline *polyline = (GMSPolyline *)[polylineInstance.mapCtrl.objects objectForKey:polylineKey];
      [polylineInstance.mapCtrl.objects removeObjectForKey:polylineKey];

      NSString *propertyId = [polylineKey stringByReplacingOccurrencesOfString:@"polyline_" withString:@"polyline_property_"];
      [polylineInstance.mapCtrl.objects removeObjectForKey:propertyId];
      polyline.map = nil;
      polyline = nil;
    }];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [polylineInstance.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
  }];

}

@end
